### PR TITLE
ApiV4: GET /channels/{channel_id}/unread

### PR DESCRIPTION
**Pull request by andreistanciu24**
Please make sure you've read the [pull request](http://docs.mattermost.com/developer/contribution-guide.html#preparing-a-pull-request) section of our [code contribution guidelines](http://docs.mattermost.com/developer/contribution-guide.html).

When filling in a section please remove the help text and the above text.

#### Summary
Get unread totals for a channel.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-5243

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [ x] Added or updated unit tests (required for all new features)
- [ x] Added API documentation (required for all new APIs)
- [ x] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

File: api4/channel.go [hunk 1]
```diff
@@ -115,6 +116,33 @@ func getChannel(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	// check permissions
+	if !app.SessionHasPermissionToChannel(c.Session, c.Params.ChannelId, model.PERMISSION_READ_CHANNEL) {
+		c.SetPermissionError(model.PERMISSION_READ_CHANNEL)
+		return
+	}
+
+	if !app.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
```
Comment (jwilander) [inline api4/channel.go]: This check is not needed, please remove

File: app/channel.go [hunk 2]
```diff
@@ -172,6 +172,28 @@ func CreateDirectChannel(userId string, otherUserId string) (*model.Channel, *mo
...
+	}
+
+	channel := result.Data.(*model.ChannelUnread)
+	var channelUnread = &model.ChannelUnread{
+		TotalMsgCount: channel.TotalMsgCount,
+		MsgCount:      0,
+		MentionCount:  channel.MentionCount,
+		TeamId:        channel.TeamId,
+		NotifyProps:   channel.NotifyProps,
+	}
```
Comment (jwilander) [inline app/channel.go]: You don't need this, you're just copying directly from one struct to another struct

File: app/channel.go [hunk 3]
```diff
@@ -172,6 +172,28 @@ func CreateDirectChannel(userId string, otherUserId string) (*model.Channel, *mo
 	}
 }
 
+func GetChannelUnread(channelId string) (*model.ChannelUnread, *model.AppError) {
+	result := <-Srv.Store.Channel().GetChannelUnread(channelId);
+	if result.Err != nil {
+		return nil, result.Err
+	}
+
+	channel := result.Data.(*model.ChannelUnread)
```
Comment (jwilander) [inline app/channel.go]: This isn't a channel so it should be called `channelUnread`

File: app/channel.go [hunk 4]
```diff
@@ -172,6 +172,28 @@ func CreateDirectChannel(userId string, otherUserId string) (*model.Channel, *mo
...
+	var channelUnread = &model.ChannelUnread{
+		TotalMsgCount: channel.TotalMsgCount,
+		MsgCount:      0,
+		MentionCount:  channel.MentionCount,
+		TeamId:        channel.TeamId,
+		NotifyProps:   channel.NotifyProps,
+	}
+
+	if channel.NotifyProps["mark_unread"] != model.CHANNEL_MARK_UNREAD_MENTION {
```
Comment (jwilander) [inline app/channel.go]: Please use `model.MARK_UNREAD_NOTIFY_PROP` constant here instead of "mark_unread"

Comment (jwilander) [inline app/channel.go]: You should probably also check the opposite case and then zero out MsgCount below instead

File: api4/channel_test.go [hunk 5]
```diff
@@ -512,6 +512,34 @@ func TestViewChannel(t *testing.T) {
...
+	defer TearDown()
+	Client := th.Client
+
+	user := th.BasicUser
+	channel := th.CreatePublicChannel()
+	app.AddUserToChannel(user, channel)
+
+	Client.Login(user.Email, user.Password)
+
+	_, resp := Client.GetChannelUnread(model.NewRandomString(16))
```
Comment (jwilander) [inline api4/channel_test.go]: Please check the success case for a normal user

File: api4/channel.go (hunk 1 above)
Comment (andreistanciu24) [inline api4/channel.go]: Done

File: app/channel.go (hunk 3 above)
Comment (andreistanciu24) [inline app/channel.go]: Done.

File: app/channel.go (hunk 2 above)
Comment (andreistanciu24) [inline app/channel.go]: I kept it this way just to control the MsgCount value (zero first), but yeah you're right. Done.

File: app/channel.go (hunk 4 above)
Comment (andreistanciu24) [inline app/channel.go]: Can you confirm that this is the correct formula for the MsgCount:
```
if channelUnread.NotifyProps[model.MARK_UNREAD_NOTIFY_PROP] != model.CHANNEL_MARK_UNREAD_MENTION {
             channelUnread.MsgCount = (channelUnread.TotalMsgCount - channelUnread.MsgCount)
	} else {
	     channelUnread.MsgCount = 0
	}
```

File: api4/channel_test.go (hunk 5 above)
Comment (andreistanciu24) [inline api4/channel_test.go]: There is already a CheckNoError test followed by checking the team id inside the channelUnread returned object.

File: api4/channel.go [hunk 6]
```diff
@@ -115,6 +116,28 @@ func getChannel(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getChannelUnread(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireChannelId()
+	if c.Err != nil {
+		return
+	}
+
+	// check permissions
```
Comment (jwilander) [inline api4/channel.go]: Can you remove this comment? It's unnecessary

File: api4/channel_test.go [hunk 7]
```diff
@@ -512,6 +512,33 @@ func TestViewChannel(t *testing.T) {
...
+	CheckBadRequestStatus(t, resp)
+	Client.Logout()
+
+	newUser := th.CreateUser()
+	Client.Login(newUser.Email, newUser.Password)
+	_, resp = Client.GetChannelUnread(th.BasicChannel.Id)
+	CheckForbiddenStatus(t, resp)
+	Client.Logout()
+
+	channelUnread, resp := th.SystemAdminClient.GetChannelUnread(channel.Id)
```
Comment (jwilander) [inline api4/channel_test.go]: Need to check the success case for a regular user as well

File: api4/channel_test.go (hunk 5 above)
Comment (jwilander) [inline api4/channel_test.go]: That is for the system admin. We need to make sure this endpoint works for regular users.

File: api4/channel.go (hunk 6 above)
Comment (andreistanciu24) [inline api4/channel.go]: Done.

File: api4/channel_test.go (hunk 7 above)
Comment (andreistanciu24) [inline api4/channel_test.go]: test case added.

File: api4/channel_test.go (hunk 5 above)
Comment (andreistanciu24) [inline api4/channel_test.go]: Done. Test case added.

**Pull request closed without merging**